src/mesh/duplicate_nodes.cc

```diff
@@ -0,0 +1,256 @@
+          MeshObject<3> te;
+          for (unsigned int i=0; i<4; i++)
+            te.nodes[i] = obj_4_node_[mesh_->node_vector.index(ele->node[i])];
+          for (unsigned int i=0; i<4; i++)
+            te.faces[i] = &triangles_[obj_4_edg_[ele->side(i)->edge_idx()]];
+          obj_4_el_.push_back(tetras_.size());
+          tetras_.push_back(te);
+        }
+        break;
+    }
+  }
+}
+
+
+/**
+ * The main functionality of the class creates duplicate nodes
+ * at interfaces with fractures.
+ * 
+ * For each node:
+ * 1) Create the lists of elements (node_elements) and edges (node_edges) sharing the node.
+ * 2) Divide node_elements into groups (components) which share one of node_edges.
+ * 3) If more than one component was created, create for each extra component a new node
+ *    and update its index in the elements from this component.
+ *    Also set node_dim_ (dimension of elements using the node).
+ */
+void DuplicateNodes::duplicate_nodes()
+{
+  FOR_NODES(mesh_, n){
+    // create list of elements
+    std::vector<unsigned int> node_elements = mesh_->node_elements[n.index()];
+    
+    // create list of edges sharing the node
+    std::set<unsigned int> node_edges;
+    for (unsigned int el_idx : node_elements)
+    {
+      FOR_ELEMENT_SIDES(&mesh_->element[el_idx], sid)
+        node_edges.insert(mesh_->element[el_idx].side(sid)->edge_idx());
+    }
+    
+    // divide node_elements into components connected by node_edges
+    std::vector<std::set<unsigned int> > components;
+    while (node_elements.size() > 0)
+    {
+      // create component containing the first element in node_elements
+      std::queue<unsigned int> q;
+      q.push(node_elements[0]);
+      node_elements.erase(node_elements.begin());
+      std::set<unsigned int> component;
+      while (q.size() > 0) {
+        auto elem = mesh_->element.full_iter(&mesh_->element[q.front()]);
+        component.insert(elem.index());
+        // add to queue adjacent elements sharing one of node_edges
+        FOR_ELEMENT_SIDES(elem, sid) {
+          auto side = elem->side(sid);
+          if (node_edges.find(side->edge_idx()) != node_edges.end())
+          {
+            unsigned int esid;
+            FOR_EDGE_SIDES(side->edge(), esid) {
+              auto adj_el_idx = side->edge()->side(esid)->element().index();
+              if (adj_el_idx != elem.index())
+              {
+                auto it = std::find(node_elements.begin(), node_elements.end(), adj_el_idx);
+                if (it != node_elements.end()) {
+                  node_elements.erase(it);
+                  q.push(adj_el_idx);
+                }
+              }
+            }
+          }
+        }
+        q.pop();
+      }
+      components.push_back(component);
+    }
+    
+    // for each component except the first one
+    // create a duplicate of the node
+    if (components.size() > 0) {
+      node_dim_[obj_4_node_[n.index()]] = mesh_->element[*components[0].begin()].dim();
```

Comment:
So from here it seems that node_dim_ gives dim for every MeshObj<0> which should be for every mesh node and this doesn't make sense. more over since later on the push_back is used, this line can simply overwrite a previous value.